icu4c/source/common/putil.cpp

```diff
@@ -1152,13 +1174,28 @@ uprv_tzname(int n)
         && uprv_strcmp(tzid, TZ_ENV_CHECK) != 0
 #endif
     ) {
-        /* The colon forces tzset() to treat the remainder as zoneinfo path */
+        /* The colon forces tzset() to treat the remainder as zoneinfo path. At
+         * least some versions of glibc does not care about the presence of a
+         * colon. So, no check is added here.*/
         if (tzid[0] == ':') {
             tzid++;
         }
-        /* This might be a good Olson ID. */
+        /* This might be a good Olson ID or it can be a full file path to a
+         * zoneinfo file or TZDEFAULT that is a symlink to a zoneinfo file or
+         * contains the actual zoneinfo rules. The last case need not be handled
+         * here. */
         skipZoneIDPrefix(&tzid);
-        return tzid;
+#if (defined(CHECK_LOCALTIME_LINK) && !defined(DEBUG_SKIP_LOCALTIME_LINK)) && (U_PLATFORM == U_PF_LINUX)
+        if (isValidOlsonID(tzid) && tzid[0] != '/')
+#endif
+          return tzid;
+#if (defined(CHECK_LOCALTIME_LINK) && !defined(DEBUG_SKIP_LOCALTIME_LINK)) && (U_PLATFORM == U_PF_LINUX)
+        /* Caller must handle threading issues */
+        else if (gTimeZoneBufferPtr != nullptr ||
+                 extractOlsonIdFromPath(tzid, uprv_strlen(tzid))) {
+          return gTimeZoneBufferPtr;
+        }
+#endif
```

Comment:
optional: I wonder if this gives the same result but might be more readable.  (FWIW, I see similarly tricky preprocessor conditions being interspersed in lines 1171-1176.)

```suggestion
#if (defined(CHECK_LOCALTIME_LINK) && !defined(DEBUG_SKIP_LOCALTIME_LINK)) && (U_PLATFORM == U_PF_LINUX)
        if (isValidOlsonID(tzid) && tzid[0] != '/') {
          return tzid;
        }
        /* Caller must handle threading issues */
        else if (gTimeZoneBufferPtr != nullptr ||
                 extractOlsonIdFromPath(tzid, uprv_strlen(tzid))) {
          return gTimeZoneBufferPtr;
        }
#else
        return tzid;
#endif
```